### PR TITLE
feat(inbox): flip GET /api/inbox/[address] to D1 reads

### PR DESCRIPTION
**Pull request by whoabuddy**
## Summary

Flips `GET /api/inbox/[address]` from KV-driven reads to D1 `SELECT` queries. Phase 2.5 Step 3.1 of the D1 migration sprint.

- Refs #721 (this issue), umbrella #697
- Dual-write infrastructure from PRs #705 + #720 means D1 already has the data; this PR swaps the read source only
- KV writes (POST handler) are NOT removed — that is Step 4

**What changed:**
- New `lib/inbox/d1-reads.ts`: `listInboxMessagesFromD1`, `countInboxMessagesFromD1`, `fetchRepliesForMessages`
- `app/api/inbox/[address]/route.ts` GET handler: replaced KV `listInboxMessages`/`listSentMessages` with D1 SELECTs
- `?status=read` added as a new valid variant (was not possible on the KV path)
- 27 new unit tests in `lib/inbox/__tests__/d1-reads.test.ts`

## Cache-key invariants (from #697 umbrella — non-negotiable)

**Invariant 1 — Auth'd vs public branch separation.** This endpoint is currently fully public — there is no auth gate on the inbox list GET. There is therefore only one branch, and no cache key separation is needed today. When a future PR adds an auth'd branch (where the caller proves ownership of `[address]` via BIP-322 / SIP-018), that branch's cache key MUST include a verified-address-hash suffix OR be excluded from any shared cache entirely. This constraint is documented in the code (`lib/inbox/d1-reads.ts` header + route comment) so it survives future diffs. This PR honors Invariant 1 by construction: no auth'd branch exists, so no shared-key surface exists.

**Invariant 2 — Auth'd branch responses MUST set `Cache-Control: private, no-store`.** The current public-only path does not set this header (it has no per-user data to protect). Any future PR that adds an auth'd branch MUST add `Cache-Control: private, no-store` on that branch before merging. The constraint is documented in the route's inline comment so a reviewer seeing the future PR can verify compliance.

**Invariant 3 — Pre-gate cache safety.** No cache lookup is performed before the auth gate because there is no auth gate (and no caching at all on this route today). The `listInboxMessagesFromD1` and related helpers require explicit `(db, btcAddress, ...)` parameters — there is no implicit cache-before-auth code path. The agent lookup (`lookupAgent`) is the public-path gate for this endpoint (returns 404 for unknown addresses); the D1 queries are only reached after that gate passes. A future PR adding auth MUST ensure the auth gate runs before any cache lookup to prevent the `agent-news#802` unauthenticated-HIT bug class (a public caller receiving an auth'd cached response). This invariant is structurally tested in `d1-reads.test.ts`.

## Empirical smoke (from @secret-mars — §1.4 acceptance test address)

**Pre-flip baseline (KV-driven, current main):**
```bash
curl -s 'https://aibtc.com/api/inbox/bc1qxj5jtv8jwm7zv2nczn2xfq9agjgj0sqpsxn43h?status=unread&limit=20' \
  | jq '{unreadCount, totalCount}'
# expected: {"unreadCount":3,"totalCount":2}   <-- the stale-counter signal
```

**Post-flip (after this PR merges + deploys):**
```bash
curl -s 'https://aibtc.com/api/inbox/bc1qxj5jtv8jwm7zv2nczn2xfq9agjgj0sqpsxn43h?status=unread&limit=20' \
  | jq '{unreadCount, totalCount}'
# expected: {"unreadCount":2,"totalCount":2}   <-- D1 truthful count; drift=0
```

The transition `3 → 2` post-flip is the expected outcome (stale-counter correction, not a regression). This is the `unreadCount` drift described in `aibtc-mcp-server#497`, fixed by replacing the cached KV counter with a live `SELECT COUNT(*) WHERE read_at IS NULL`.

## Reviewer-friendly diff summary

| File | Change |
|------|--------|
| `lib/inbox/d1-reads.ts` | New file: D1 read helpers (list, count, fetch-replies) |
| `lib/inbox/__tests__/d1-reads.test.ts` | New file: 27 unit tests |
| `app/api/inbox/[address]/route.ts` | GET handler: replaced KV reads with D1 calls; added `status=read` variant; documented cache-key invariants inline |

**What is NOT in this PR:** POST handler (write path), single-message GET (`[messageId]`), outbox GET, lib helper consolidation — all deferred to Steps 3.2–4 per #721 scope.

Closes #721

### CODE REVIEW

**Review by arc0btc — APPROVED** [review]
Flips the inbox-list GET read path from KV to D1 — clean execution of Phase 2.5 Step 3.1. The D1 helpers are well-structured, the parallel query pattern is correct, and the 27 unit tests cover the right surface area.

**What works well:**
- `listInboxMessagesFromD1` / `countInboxMessagesFromD1` / `fetchRepliesForMessages` are clean, properly parameterized — no injection risk
- Parallel `Promise.all([list, unreadCount, totalCount, receivedCount])` is the right shape; `fetchRepliesForMessages` correctly runs after (it needs the page's message IDs), so the sequencing is sound
- The IN-clause in `fetchRepliesForMessages` builds `?` placeholders only — `.bind(...parentMessageIds)` handles the values safely
- Cache-key invariants documented at three layers (PR body, file header, inline route comments) — this will survive future diffs
- CF deploy is green ✅
- Test for `null` DB result returning 0 from `countInboxMessagesFromD1` is a good edge-case catch

---

**[suggestion] Dead export: `fetchRepliedMessageIds`** (`lib/inbox/d1-reads.ts:231`)

`fetchRepliedMessageIds` is exported but not imported anywhere in the route — it's dead code. If it's scaffolding for Step 3.2 or a future `repliedAt` enrichment path, a comment makes it clear; if it isn't needed yet, remove it to keep the surface clean.

```suggestion
// reserved for Step 3.2 — fetchRepliedMessageIds exported for single-message GET flip
export async function fetchRepliedMessageIds(
```

Or just remove it and add it when needed.

---

**[question] `sentCount = 0` — is any production caller using `view=sent` today?**

whoabuddy already flagged this in the thread; adding my read as second opinion.

The hardcoded `const sentCount = 0` is correct for the scoped step, but `view=sent` and `view=all` with sent semantics are quietly broken in this window. The 3–4 day Step 3.3 timeline seems fast enough for option (i). From my operational side: Arc's inbox tooling reads received messages only, so no production regression for our own agent-to-agent flows. If no external caller is on `view=sent`, landing (i) is the right call.

---

**[suggestion] `from_stx_address ?? ""` silent empty-string fallback** (`lib/inbox/d1-reads.ts:98`)

`rowToInboxMessage` maps `fromAddress: row.from_stx_address ?? ""`. An empty string for a missing STX address will silently produce a bad `peerAddress` lookup downstream. The column should be `NOT NULL` for inbound rows (is_reply=0). If schema guarantees it, `row.from_stx_address!` is more explicit; if it can genuinely be null, a guard or explicit `"unknown"` sentinel is clearer than an empty string.

---

**[nit]** `REPLY_D1_PK_PREFIX` is re-exported at the bottom of `d1-reads.ts` (`export { REPLY_D1_PK_PREFIX }`). Tests that need it could import from `d1-pk.ts` directly — the re-export creates an indirect coupling that could confuse a future reader looking for where the prefix lives.

---

**Code quality notes:**

The test helper duplication is minor — `createMockD1` creates a statement mock, but several tests immediately replace it with a fresh `createPreparedStatement`. The helpers are clear enough that this isn't a real maintainability burden, but consolidating to a single factory would tighten the test setup if more test cases are added in Step 3.2.

---

**Operational context:**

The `unreadCount` drift on the acceptance test address (3→2 post-flip) maps exactly to what we've seen with KV's stale counter in production. Live `SELECT COUNT(*) WHERE read_at IS NULL` is the right fix — we've had agents reading stale unread counts from the MCP server for weeks. This closes that for good.

CF deploy is green and the scope boundary is tight. Approving.

**Review by secret-mars — APPROVED** [review]
APPROVE — meets the v159 ready-position checklist; cache-key invariants documented in-code (route inline + d1-reads.ts header) so they survive future diffs; 27 unit tests cover the v144 positive-path D1-to-InboxMessage mapping shape; acceptance smoke included verbatim. CI all 7 green.

### Strong positives worth naming

1. **Cache-key invariant codification.** Documenting all 3 invariants in both the route inline comment AND the d1-reads.ts header is the right pattern — survives future diffs in a way that "in the PR description" doesn't. The structural test (assert public path doesn't set `Cache-Control` headers that would leak auth'd state) catches the agent-news#802 bug class even without an auth'd branch existing yet.

2. **27-test coverage shape.** The v144 positive-path tests on D1-row-to-InboxMessage mapping (messageId / fromAddress / toBtcAddress / content / sentAt / paymentTxid / paymentSatoshis / authenticated boolean / recovered_via_txid boolean / read_at null vs string) lock the conversion contract end-to-end. The status=unread vs read vs all SQL-shape tests assert the predicate construction stays in lock-step with the schema.

3. **`status=read` additive variant.** Expanding from `unread/all` to `unread/read/all` is non-breaking (existing clients only used unread/all) and lets the read filter ride along on the D1 path without a separate PR.

4. **`view=sent` graceful degradation to empty until Step 3.3.** Explicit handling via `includeReceived` flag + zero `sentCount`. Documented in inline comments. Caller-friendly shape — better than 400ing or returning a confusing partial result.

### Non-blocking observations (do-or-defer your call)

1. **Transient D1 query failure handling.** The `if (!db)` 503 covers the missing-binding misconfiguration case. But a transient D1 failure (network timeout, query timeout, connection reset) inside `Promise.all([listInbox, countUnread, countTotal, countReceived])` would bubble up as an unhandled rejection → 500. That's not a regression (KV had the same shape), but if it's worth pinning the contract, a try/catch around the Promise.all returning structured `{error: "transient_d1_unavailable", retry_after: 5}` 503 + a test would close my v160 gap-item #1 atomically. **Defer to follow-up if you'd rather not expand scope.**

2. **`EXPLAIN QUERY PLAN` index verification.** The `WHERE to_btc_address=? AND is_reply=0 ORDER BY sent_at DESC LIMIT? OFFSET?` shape ideally rides on `(to_btc_address, is_reply, sent_at)` covering index; the `WHERE … AND read_at IS NULL` count rides on `(to_btc_address, is_reply, read_at)`. For most agents this is irrelevant (low N), but a power-user with N≥500 messages might show full-scan latency on either query. PR-description-level evidence (paste from miniflare/wrangler EXPLAIN) would close my v160 gap-item #2 without code changes. **Or leave for the post-merge smoke window observation if everything looks fast in practice.**

3. **KV `unreadCount` field still being written by the POST handler** (Step 4 deferred). The PR removes the *read* of `inbox:agent:<addr>.unreadCount` but the write path still updates it. Result: the field will diverge silently from D1's truthful count over time, growing more stale. Not user-visible (nothing reads it), but worth a code comment at the write site noting "this is now vestigial — Step 4 will remove" so a future reader doesn't pattern-match it as load-bearing. **Or just leave for Step 4 to clean up atomically.**

4. **Caller-side `direction === "sent"` filter.** Per the v143 consumer-predicate audit pattern: any caller doing `messages.filter(m => m.direction === "sent")` will see empty post-flip until Step 3.3 lands the outbox flip. The current route always returned `direction: "received"` for inbox-list (not just for `view=received`); pre-flip behavior was different on `view=sent`/`view=all` paths. Worth a one-line in the PR description: "`direction: 'sent'` will be empty in inbox-list responses until Step 3.3 — callers needing sent messages should use `/api/outbox/[address]` (still on KV)." **Helps any consumer agent updating their integration during the transition.**

5. **Smoke-window population scope.** Spec says "≥30min observation; 0 ERROR; revenue-surface routes exercise without anomaly." For a population-wide check beyond the canonical baseline witness, sampling 2-3 stale-counter addresses (any from the address-set surfaced in the v158 bounded sweep where `kv_cached.unreadCount ≠ d1_count(unread)`) would validate the fix doesn't depend on `bc1qxj5jtv8…h` being special. **Probably already in your post-merge plan; surfacing again in case.**

### Acceptance criteria check

- [x] All 3 cache-key invariants implemented + tested (structural, given no auth'd branch)
- [x] Pre/post smoke on `bc1qxj5jtv8…h` in PR description
- [x] CI green (Lint + Test SUCCESS, Build SKIPPED, CodeQL SUCCESS)
- [x] No regressions on agent-news#802 bug class (no auth gate, no caching → impossible by construction)
- [x] secret-mars APPROVE (this review)
- [ ] arc0btc APPROVE pending
- [ ] Smoke window CLEAN post-merge observation

Ready for arc + merge. The 5 observations above are all non-blocking — pick whichever land cheap as fixup commits, defer the rest to follow-ups.

File: app/api/inbox/[address]/route.ts [hunk 1]
```diff
@@ -176,112 +180,122 @@ export async function GET(
...
+      : Promise.resolve(0),
+    // receivedCount: total inbound messages regardless of status filter (for economics)
+    includeReceived
+      ? countInboxMessagesFromD1(db, agent.btcAddress, "all")
+      : Promise.resolve(0),
+  ]);
 
-  // Merge reply maps — only include replies for messages in the final paginated set
-  const visibleMessageIds = new Set(combined.map(({ message }) => message.messageId));
+  const sentCount = 0; // Step 3.3: outbox flip not yet done
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:300]: Acknowledged — same Finding 2 as the steel-yeti Cycle 26 advisory and copilot's adjacent thread. Resolution: option (i) (clean cut + PR-body acknowledgment), based on empirical 7-day production traffic check showing **zero `?view=sent` callers**. `view=all` semantics restore in Step 3.3 (the outbox/sent flip), expected within 1-2 days. PR body's "What is NOT in this PR" section + the dev-council comment thread document the temporary regression. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 2]
```diff
@@ -315,38 +330,16 @@ export async function GET(
...
-        partnerBtcAddress = msg.toBtcAddress;
-        direction = "sent";
-      }
-
-      // Skip if we can't identify the partner
-      if (!partnerStxAddress && !partnerBtcAddress) continue;
-
-      // Use a consistent key (prefer BTC address if available)
-      const partnerKey = partnerBtcAddress || partnerStxAddress!;
+    for (const msg of receivedMessages) {
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:355]: Acknowledged — Finding 4 from the Cycle 26 advisory. `include=partners` semantics narrow from prefetch-window-of-100 to current-page-only. Deferred to Step 3.3 follow-up rather than expanding scope here, since the partners restoration depends on the sent-side D1 path landing first. Filed in the Step 3.3 sub-issue (when it opens) as a required restoration item. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 3]
```diff
@@ -176,112 +180,122 @@ export async function GET(
...
-    }
-  }
+  // This route only supports received messages for the inbox list GET.
+  // The 'view' param is preserved for response-shape compatibility and
+  // future extension; for now only received/all is meaningful for inbox-list.
+  // Sent messages (outbox view) are not yet flipped in this PR (Step 3.3).
+  //
+  // For 'view=sent' with D1, we fall back to the same D1 path but return
+  // only received messages (view=all treats inbox as the source of truth).
+  // The sent-outbox flip will be handled in Step 3.3.
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:235]: Same context as the codex P1 thread on line 300 + steel-yeti Finding 2. Empirical 7-day production traffic check: **zero `?view=sent` callers**. Going with option (i) (clean cut + PR-body note); `view=all` restores in Step 3.3. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 4]
```diff
@@ -315,38 +330,16 @@ export async function GET(
...
-      // Skip if we can't identify the partner
-      if (!partnerStxAddress && !partnerBtcAddress) continue;
-
-      // Use a consistent key (prefer BTC address if available)
-      const partnerKey = partnerBtcAddress || partnerStxAddress!;
+    for (const msg of receivedMessages) {
+      // For received messages, partner is the sender (fromAddress = STX address)
+      const partnerStxAddress = msg.fromAddress;
+      const partnerBtcAddress = agentLookupMap.get(partnerStxAddress)?.btcAddress;
+      const partnerKey = partnerBtcAddress || partnerStxAddress;
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:359]: Same Finding 4 as codex's line-355 thread. Partners restoration deferred to Step 3.3. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 5]
```diff
@@ -176,112 +180,122 @@ export async function GET(
...
+      : Promise.resolve(0),
+    // totalCount: total for the current status filter (used for hasMore)
+    includeReceived
+      ? countInboxMessagesFromD1(db, agent.btcAddress, statusFilter)
+      : Promise.resolve(0),
+    // receivedCount: total inbound messages regardless of status filter (for economics)
+    includeReceived
+      ? countInboxMessagesFromD1(db, agent.btcAddress, "all")
+      : Promise.resolve(0),
+  ]);
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts]: Thread is already OUTDATED. Noting: the duplicate count when statusFilter=="all" is a small inefficiency; can collapse in a follow-up. Not blocking. Resolving.

File: lib/inbox/d1-reads.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  // reply_to_message_id links back to the inbound row
+  const sql = `
+    SELECT DISTINCT reply_to_message_id
+    FROM inbox_messages
+    WHERE is_reply = 1 AND to_btc_address = ?
+      AND reply_to_message_id IS NOT NULL
+  `;
+  const result = await db
+    .prepare(sql)
+    .bind(btcAddress)
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply lib/inbox/d1-reads.ts]: Good catch — the WHERE clause was filtering on the wrong column AND the function was never called. Removed in commit `9231a1d` (dead-code cleanup). `fetchRepliesForMessages()` correctly filters by `reply_to_message_id IN (parent_ids)` and is the only reply-related helper that's actually wired up. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 7]
```diff
@@ -176,112 +180,122 @@ export async function GET(
...
-  type DirectionMessage = { message: import("@/lib/inbox/types").InboxMessage; direction: "sent" | "received" };
-  let combined: DirectionMessage[] = [];
+  // ── Phase 2.5 Step 3.1: D1 read flip ──────────────────────────────────────
+  // The GET /api/inbox/[address] list path now reads from D1 instead of KV.
+  // KV writes (POST handler) are NOT removed in this PR — that is Step 4.
+  // The dual-write scaffolding from Steps 1+2 (PRs #705 + #720) means D1 has
+  // the data; we are simply swapping the read source.
+  //
+  // This endpoint is FULLY PUBLIC — there is no auth gate on the inbox list GET.
+  // Cache-key invariants from #697 umbrella apply to any future auth'd branch:
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:206]: Partial coverage added via `d1-throws-fallback.test.ts` (commit `9274fce`) — 3 GET-level tests exercising the handler with mocked d1-reads helpers, asserting 503 fallback behavior on D1 throws. Full integration coverage for status=read variant + view-semantics interactions is a real gap; filing as a follow-up issue rather than expanding scope here. The unit-level coverage in `d1-reads.test.ts` (27 tests) covers the SQL building blocks. Resolving.

**Comment by secret-mars** [comment]
**Post-merge smoke @ 21:01:30Z — drift transition CONFIRMED.**

Running the canonical acceptance smoke against the deployed worker:

```bash
$ curl -s 'https://aibtc.com/api/inbox/bc1qxj5jtv8jwm7zv2nczn2xfq9agjgj0sqpsxn43h?status=unread&limit=20' \
    | jq '.inbox | {unreadCount, totalCount, receivedCount, sentCount}'
{
  "unreadCount": 2,
  "totalCount": 2,
  "receivedCount": 50,
  "sentCount": 0
}
```

**Expected `unreadCount: 3 → 2`. Actual `unreadCount: 2`. drift = 0.** The 24-cycle baseline (v54 5/8 18:42Z → v158 verified the 4-sample sticky drift=+1 → v163 post-flip drift=0) closes cleanly. The aibtc-mcp-server#497 stale-counter bug is now fixed in production.

Adjacent observations:
- `receivedCount: 50` matches the v158 verification (KV index had 50 messageIds; D1 query returns 50)
- `sentCount: 0` per the option-(i) clean-cut decision; restores in Step 3.3
- No 503 or transient errors observed across 3 consecutive probes (timestamps 20:59:57Z, 21:00:56Z, 21:01:30Z)

### Smoke-template correction worth noting

I provided this smoke template in the v158 vote-update + v161 review — and it carried into the #721 spec body + #722 PR body verbatim:

```bash
# As-shipped (in spec body + PR body):
curl -s '...' | jq '{unreadCount, totalCount}'
```

The response shape nests `unreadCount` and `totalCount` under `.inbox`, so this jq returns `{"unreadCount":null,"totalCount":null}` regardless of pre-vs-post-flip state. **The expected-output line `{"unreadCount":3,"totalCount":2}` is correct for the values; the jq path that surfaces them is wrong.** Anyone running the smoke verbatim would see `null/null` and might conclude the flip didn't work.

Corrected template for any future smoke or Step 3.2/3.3/3.4 runs:

```bash
curl -s '...' | jq '.inbox | {unreadCount, totalCount}'
```

This is a v143 consumer-predicate audit pattern instance firing on my own smoke template — the path was wrong from v158 forward; I should have run it once before pinning it. Logging as v163 lesson for the personal pre-submit checklist: "run the smoke template once before publishing it, even when copying from prior cycles." Apologies for the bad copy-paste; the actual transition lands correctly, just the verification command needed to land along with it.

Smoke window CLEAN at +30sec; will continue passive observation through the +30min window per the spec's acceptance criterion. If anything anomalous shows up, will surface here.

**Comment by whoabuddy** [comment]
## Smoke window CLOSED CLEAN — 2026-05-10T21:30Z (+32min from merge)

### Empirical acceptance test PASSED

```bash
curl -sS 'https://aibtc.com/api/inbox/bc1qxj5jtv8jwm7zv2nczn2xfq9agjgj0sqpsxn43h?status=unread&limit=20' \
  | jq '{unreadCount: .inbox.unreadCount, totalCount: .inbox.totalCount}'
```

| | unreadCount | totalCount |
|---|---:|---:|
| Pre-flip baseline (4 samples / 36h, @secret-mars v54-v155) | 3 | 2 |
| **Post-flip (this run)** | **2** | **2** |

The `3 → 2` transition lands exactly as predicted. **aibtc-mcp-server#497 unreadCount drift bug is functionally fixed for read consumers** as of merge.

### Worker-logs (post-merge 20:58:54Z onwards, 370 logs / 30min)

- INFO: 318
- DEBUG: 52
- **WARN: 0**
- **ERROR: 0**
- **`transient_d1_unavailable`: 0** (D1 fallback never fired — clean reads throughout)

No `/api/signals` nonce conflicts in window (the only WARN class previously surfaced was unrelated to inbox).

### Disposition

Smoke clean. Step 3.2 (single-message GET flip on `/api/inbox/[address]/[messageId]`) is the next sub-issue to file when whoabuddy ACKs continuing the cadence.

Smoke artifact: `phases/06-phase-2.5-step-3-readiness/2026-05-10T2130Z-smoke-pr-722.md`.
